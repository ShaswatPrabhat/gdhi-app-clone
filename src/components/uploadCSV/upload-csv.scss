@import "../../assets/stylesheets/components/variables";

.error-message {
  color: $negative;
}

.button-section {
  .upload-file {
    display: none;
  }
  .btn {
    margin-left: 0px;
    margin-right: 16px;
  }
}
