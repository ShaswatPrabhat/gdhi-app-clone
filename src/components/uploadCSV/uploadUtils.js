import { object, string, number, boolean } from "yup";

const csvDataMap = {
  CountryName: "Country Name",
  CountrySummary: "Country Summary",
  CountryContactName: "Country Contact Name",
  CountryContactRole: "Country Contact Role",
  CountryContactOrg: "Country Contact Org",
  CountryContactEmail: "Country Contact Email",
  ContactPersonName: "Contact Person Name",
  ContactPersonRole: "Contact Person Role",
  ContactPersonEmail: "Contact Person Email",
  IsTheDataApprovedByTheGovernment: "Is the data approved by the government",
  DataApproverName: "Data Approver Name",
  DataApproverRole: "Data Approver Role",
  DataApproverEmail: "Data Approver Email",
  ResourcesLink: "Resources Link",
  Indicator1Score: "Indicator 1 Score",
  EnterIndicator1Justification: "Enter Indicator 1 justification",
  Indicator2Score: "Indicator 2 Score",
  EnterIndicator2Justification: "Enter Indicator 2 justification",
  Indicator2aScore: "Indicator 2a Score",
  EnterIndicator2aJustification: "Enter Indicator 2a justification",
  Indicator3Score: "Indicator 3 Score",
  EnterIndicator3Justification: "Enter Indicator 3 justification",
  Indicator4Score: "Indicator 4 Score",
  EnterIndicator4Justification: "Enter Indicator 4 justification",
  Indicator4aScore: "Indicator 4a Score",
  EnterIndicator4aJustification: "Enter Indicator 4a justification",
  Indicator5Score: "Indicator 5 Score",
  EnterIndicator5Justification: "Enter Indicator 5 justification",
  Indicator5aScore: "Indicator 5a Score",
  EnterIndicator5aJustification: "Enter Indicator 5a justification",
  Indicator6Score: "Indicator 6 Score",
  EnterIndicator6Justification: "Enter Indicator 6 justification",
  Indicator6aScore: "Indicator 6a Score",
  EnterIndicator6aJustification: "Enter Indicator 6a justification",
  Indicator7Score: "Indicator 7 Score",
  EnterIndicator7Justification: "Enter Indicator 7 justification",
  Indicator8Score: "Indicator 8 Score",
  EnterIndicator8Justification: "Enter Indicator 8 justification",
  Indicator9Score: "Indicator 9 Score",
  EnterIndicator9Justification: "Enter Indicator 9 justification",
  Indicator9aScore: "Indicator 9a Score",
  EnterIndicator9aJustification: "Enter Indicator 9a justification",
  Indicator10Score: "Indicator 10 Score",
  EnterIndicator10Justification: "Enter Indicator 10 justification",
  Indicator11Score: "Indicator 11 Score",
  EnterIndicator11Justification: "Enter Indicator 11 justification",
  Indicator12Score: "Indicator 12 Score",
  EnterIndicator12Justification: "Enter Indicator 12 justification",
  Indicator13Score: "Indicator 13 Score",
  EnterIndicator13Justification: "Enter Indicator 13 justification",
  Indicator14Score: "Indicator 14 Score",
  EnterIndicator14Justification: "Enter Indicator 14 justification",
  Indicator15Score: "Indicator 15 Score",
  EnterIndicator15Justification: "Enter Indicator 15 justification",
  Indicator16Score: "Indicator 16 Score",
  EnterIndicator16Justification: "Enter Indicator 16 justification",
  Indicator17Score: "Indicator 17 Score",
  EnterIndicator17Justification: "Enter Indicator 17 justification",
  Indicator18Score: "Indicator 18 Score",
  EnterIndicator18Justification: "Enter Indicator 18 justification",
  Indicator19Score: "Indicator 19 Score",
  EnterIndicator19Justification: "Enter Indicator 19 justification",
  Indicator20Score: "Indicator 20 Score",
  EnterIndicator20Justification: "Enter Indicator 20 justification",
  Indicator21Score: "Indicator 21 Score",
  EnterIndicator21Justification: "Enter Indicator 21 justification",
  Indicator21aScore: "Indicator 21a Score",
  EnterIndicator21aJustification: "Enter Indicator 21a justification",
  Indicator21bScore: "Indicator 21b Score",
  EnterIndicator21bJustification: "Enter Indicator 21b justification",
  Indicator21cScore: "Indicator 21c Score",
  EnterIndicator21cJustification: "Enter Indicator 21c justification",
  Indicator22Score: "Indicator 22 Score",
  EnterIndicator22Justification: "Enter Indicator 22 justification",
  Indicator23Score: "Indicator 23 Score",
  EnterIndicator23Justification: "Enter Indicator 23 justification",
};

export const generatePayloadFromParsedJSON = (formCSVRow) => ({
  countryId: "",
  currentYear: "",
  dataAvailableForYear: "",
  status: "",
  updatedDate: "",
  countrySummary: {
    countryId: "",
    countryName: formCSVRow[csvDataMap.CountryName],
    countryAlpha2Code: "",
    summary: formCSVRow[csvDataMap.CountrySummary],
    contactName: formCSVRow[csvDataMap.CountryContactName],
    contactDesignation: formCSVRow[csvDataMap.CountryContactRole],
    contactOrganization: formCSVRow[csvDataMap.CountryContactOrg],
    contactEmail: formCSVRow[csvDataMap.CountryContactEmail],
    dataFeederName: formCSVRow[csvDataMap.ContactPersonName],
    dataFeederRole: formCSVRow[csvDataMap.ContactPersonRole],
    dataFeederEmail: formCSVRow[csvDataMap.ContactPersonEmail],
    dataApproverName: formCSVRow[csvDataMap.DataApproverName],
    dataApproverRole: formCSVRow[csvDataMap.DataApproverRole],
    dataApproverEmail: formCSVRow[csvDataMap.DataApproverEmail],
    govtApproved: formCSVRow[csvDataMap.IsTheDataApprovedByTheGovernment],
    resources: formCSVRow[csvDataMap.ResourcesLink]
      .split(",")
      .map(function (resourceLink) {
        return resourceLink.trim();
      }),
  },
  healthIndicators: [
    {
      categoryId: 1,
      indicatorId: 1,
      status: "",
      score: formCSVRow[csvDataMap.Indicator1Score],
      supportingText: formCSVRow[csvDataMap.EnterIndicator1Justification],
    },
    {
      categoryId: 1,
      indicatorId: 2,
      status: "",
      score: formCSVRow[csvDataMap.Indicator2Score],
      supportingText: formCSVRow[csvDataMap.EnterIndicator2Justification],
    },
    {
      categoryId: 1,
      indicatorId: 30,
      status: "",
      score: formCSVRow[csvDataMap.Indicator2aScore],
      supportingText: formCSVRow[csvDataMap.EnterIndicator2aJustification],
    },
    {
      categoryId: 1,
      indicatorId: 31,
      status: "",
      score: formCSVRow[csvDataMap.Indicator3Score],
      supportingText: formCSVRow[csvDataMap.EnterIndicator3Justification],
    },
    {
      categoryId: 1,
      indicatorId: 32,
      status: "",
      score: formCSVRow[csvDataMap.Indicator4Score],
      supportingText: formCSVRow[csvDataMap.EnterIndicator4Justification],
    },
    {
      categoryId: 1,
      indicatorId: 33,
      status: "",
      score: formCSVRow[csvDataMap.Indicator4aScore],
      supportingText: formCSVRow[csvDataMap.EnterIndicator4aJustification],
    },
    {
      categoryId: 2,
      indicatorId: 3,
      status: "",
      score: formCSVRow[csvDataMap.Indicator5Score],
      supportingText: formCSVRow[csvDataMap.EnterIndicator5Justification],
    },
    {
      categoryId: 2,
      indicatorId: 37,
      status: "",
      score: formCSVRow[csvDataMap.Indicator5aScore],
      supportingText: formCSVRow[csvDataMap.EnterIndicator5aJustification],
    },
    {
      categoryId: 2,
      indicatorId: 4,
      status: "",
      score: formCSVRow[csvDataMap.Indicator6Score],
      supportingText: formCSVRow[csvDataMap.EnterIndicator6Justification],
    },
    {
      categoryId: 2,
      indicatorId: 38,
      status: "",
      score: formCSVRow[csvDataMap.Indicator6aScore],
      supportingText: formCSVRow[csvDataMap.EnterIndicator6aJustification],
    },
    {
      categoryId: 3,
      indicatorId: 5,
      status: "",
      score: formCSVRow[csvDataMap.Indicator7Score],
      supportingText: formCSVRow[csvDataMap.EnterIndicator7Justification],
    },
    {
      categoryId: 3,
      indicatorId: 6,
      status: "",
      score: formCSVRow[csvDataMap.Indicator8Score],
      supportingText: formCSVRow[csvDataMap.EnterIndicator8Justification],
    },
    {
      categoryId: 3,
      indicatorId: 7,
      status: "",
      score: formCSVRow[csvDataMap.Indicator9Score],
      supportingText: formCSVRow[csvDataMap.EnterIndicator9Justification],
    },
    {
      categoryId: 3,
      indicatorId: 36,
      status: "",
      score: formCSVRow[csvDataMap.Indicator9aScore],
      supportingText: formCSVRow[csvDataMap.EnterIndicator9aJustification],
    },
    {
      categoryId: 3,
      indicatorId: 8,
      status: "",
      score: formCSVRow[csvDataMap.Indicator10Score],
      supportingText: formCSVRow[csvDataMap.EnterIndicator10Justification],
    },
    {
      categoryId: 4,
      indicatorId: 9,
      status: "",
      score: formCSVRow[csvDataMap.Indicator11Score],
      supportingText: formCSVRow[csvDataMap.EnterIndicator11Justification],
    },
    {
      categoryId: 4,
      indicatorId: 10,
      status: "",
      score: formCSVRow[csvDataMap.Indicator12Score],
      supportingText: formCSVRow[csvDataMap.EnterIndicator12Justification],
    },
    {
      categoryId: 4,
      indicatorId: 11,
      status: "",
      score: formCSVRow[csvDataMap.Indicator13Score],
      supportingText: formCSVRow[csvDataMap.EnterIndicator13Justification],
    },
    {
      categoryId: 4,
      indicatorId: 12,
      status: "",
      score: formCSVRow[csvDataMap.Indicator14Score],
      supportingText: formCSVRow[csvDataMap.EnterIndicator14Justification],
    },
    {
      categoryId: 5,
      indicatorId: 13,
      status: "",
      score: formCSVRow[csvDataMap.Indicator15Score],
      supportingText: formCSVRow[csvDataMap.EnterIndicator15Justification],
    },
    {
      categoryId: 5,
      indicatorId: 14,
      status: "",
      score: formCSVRow[csvDataMap.Indicator16Score],
      supportingText: formCSVRow[csvDataMap.EnterIndicator16Justification],
    },
    {
      categoryId: 6,
      indicatorId: 15,
      status: "",
      score: formCSVRow[csvDataMap.Indicator17Score],
      supportingText: formCSVRow[csvDataMap.EnterIndicator17Justification],
    },
    {
      categoryId: 6,
      indicatorId: 16,
      status: "",
      score: formCSVRow[csvDataMap.Indicator18Score],
      supportingText: formCSVRow[csvDataMap.EnterIndicator18Justification],
    },
    {
      categoryId: 7,
      indicatorId: 17,
      status: "",
      score: formCSVRow[csvDataMap.Indicator19Score],
      supportingText: formCSVRow[csvDataMap.EnterIndicator19Justification],
    },
    {
      categoryId: 7,
      indicatorId: 18,
      status: "",
      score: formCSVRow[csvDataMap.Indicator20Score],
      supportingText: formCSVRow[csvDataMap.EnterIndicator20Justification],
    },
    {
      categoryId: 7,
      indicatorId: 19,
      status: "",
      score: formCSVRow[csvDataMap.Indicator21Score],
      supportingText: formCSVRow[csvDataMap.EnterIndicator21Justification],
    },
    {
      categoryId: 7,
      indicatorId: 27,
      status: "",
      score: formCSVRow[csvDataMap.Indicator21aScore],
      supportingText: formCSVRow[csvDataMap.EnterIndicator21aJustification],
    },
    {
      categoryId: 7,
      indicatorId: 28,
      status: "",
      score: formCSVRow[csvDataMap.Indicator21bScore],
      supportingText: formCSVRow[csvDataMap.EnterIndicator21bJustification],
    },
    {
      categoryId: 7,
      indicatorId: 29,
      status: "",
      score: formCSVRow[csvDataMap.Indicator21cScore],
      supportingText: formCSVRow[csvDataMap.EnterIndicator21cJustification],
    },
    {
      categoryId: 7,
      indicatorId: 34,
      status: "",
      score: formCSVRow[csvDataMap.Indicator22Score],
      supportingText: formCSVRow[csvDataMap.EnterIndicator22Justification],
    },
    {
      categoryId: 7,
      indicatorId: 35,
      status: "",
      score: formCSVRow[csvDataMap.Indicator23Score],
      supportingText: formCSVRow[csvDataMap.EnterIndicator23Justification],
    },
  ],
});

export async function validateFields(data) {
  const min1Max5RequiredInteger = number()
    .required()
    .transform((originalValue) => {
      if (
        originalValue === 1 ||
        originalValue === 2 ||
        originalValue === 3 ||
        originalValue === 4 ||
        originalValue === 5
      ) {
        return Number(originalValue);
      }
      return -1;
    });
  const requiredString = string().trim().required();
  const requiredEmail = string().trim().email().required();
  const email = string().trim().email();
  const trimString = string().trim();
  return await object({
    [csvDataMap.CountryName]: requiredString,
    [csvDataMap.CountrySummary]: requiredString,
    [csvDataMap.CountryContactName]: trimString,
    [csvDataMap.CountryContactRole]: trimString,
    [csvDataMap.CountryContactOrg]: trimString,
    [csvDataMap.CountryContactEmail]: email,
    [csvDataMap.ContactPersonName]: requiredString,
    [csvDataMap.ContactPersonRole]: requiredString,
    [csvDataMap.ContactPersonEmail]: requiredEmail,
    [csvDataMap.IsTheDataApprovedByTheGovernment]: boolean(),
    [csvDataMap.DataApproverName]: string()
      .trim()
      .when([csvDataMap.IsTheDataApprovedByTheGovernment], {
        is: true,
        then: () => requiredString,
      }),
    [csvDataMap.DataApproverRole]: string()
      .trim()
      .when([csvDataMap.IsTheDataApprovedByTheGovernment], {
        is: true,
        then: () => requiredString,
      }),
    [csvDataMap.DataApproverEmail]: string()
      .trim()
      .when([csvDataMap.IsTheDataApprovedByTheGovernment], {
        is: true,
        then: () => requiredEmail,
      }),
    [csvDataMap.ResourcesLink]: string()
      .trim()
      .lowercase()
      .required()
      .test(
        "validate-csv-urls",
        (d) => `${d.value} are not valid URL's`,
        function (value) {
          let isValid = true;
          value.split(",").forEach((a) => {
            const urlPattern = new RegExp(
              "^(https?:\\/\\/)?" + // validate protocol
                "((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|" + // validate domain name
                "((\\d{1,3}\\.){3}\\d{1,3}))" + // validate OR ip (v4) address
                "(\\:\\d+)?(\\/[-a-z\\d%_.~+]*)*" + // validate port and path
                "(\\?[;&a-z\\d%_.~+=-]*)?" + // validate query string
                "(\\#[-a-z\\d_]*)?$",
              "i"
            ); // validate fragment locator

            isValid = isValid && !!urlPattern.test(a.trim());
          });
          return isValid;
        }
      ),
    [csvDataMap.Indicator1Score]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator1Justification]: requiredString,
    [csvDataMap.Indicator2Score]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator2Justification]: requiredString,
    [csvDataMap.Indicator2aScore]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator2aJustification]: requiredString,
    [csvDataMap.Indicator3Score]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator3Justification]: requiredString,
    [csvDataMap.Indicator4Score]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator4Justification]: requiredString,
    [csvDataMap.Indicator4aScore]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator4aJustification]: requiredString,
    [csvDataMap.Indicator5Score]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator5Justification]: requiredString,
    [csvDataMap.Indicator5aScore]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator5aJustification]: requiredString,
    [csvDataMap.Indicator6Score]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator6Justification]: requiredString,
    [csvDataMap.Indicator6aScore]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator6aJustification]: requiredString,
    [csvDataMap.Indicator7Score]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator7Justification]: requiredString,
    [csvDataMap.Indicator8Score]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator8Justification]: requiredString,
    [csvDataMap.Indicator9Score]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator9Justification]: requiredString,
    [csvDataMap.Indicator9aScore]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator9aJustification]: requiredString,
    [csvDataMap.Indicator10Score]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator10Justification]: requiredString,
    [csvDataMap.Indicator11Score]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator11Justification]: requiredString,
    [csvDataMap.Indicator12Score]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator12Justification]: requiredString,
    [csvDataMap.Indicator13Score]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator13Justification]: requiredString,
    [csvDataMap.Indicator14Score]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator14Justification]: requiredString,
    [csvDataMap.Indicator15Score]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator15Justification]: requiredString,
    [csvDataMap.Indicator16Score]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator16Justification]: requiredString,
    [csvDataMap.Indicator17Score]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator17Justification]: requiredString,
    [csvDataMap.Indicator18Score]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator18Justification]: requiredString,
    [csvDataMap.Indicator19Score]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator19Justification]: requiredString,
    [csvDataMap.Indicator20Score]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator20Justification]: requiredString,
    [csvDataMap.Indicator21Score]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator21Justification]: requiredString,
    [csvDataMap.Indicator21aScore]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator21aJustification]: requiredString,
    [csvDataMap.Indicator21bScore]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator21bJustification]: requiredString,
    [csvDataMap.Indicator21cScore]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator21cJustification]: requiredString,
    [csvDataMap.Indicator22Score]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator22Justification]: requiredString,
    [csvDataMap.Indicator23Score]: min1Max5RequiredInteger,
    [csvDataMap.EnterIndicator23Justification]: requiredString,
  }).validate(data);
}
