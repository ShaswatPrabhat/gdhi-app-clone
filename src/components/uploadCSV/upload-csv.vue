<script>
import Vue from "vue";
import Papa from "papaparse";
import { generatePayloadFromParsedJSON, validateFields } from "./uploadUtils";
import common from "../../common/common";
import axios from "axios";

const status = Object.freeze({
  VALID: "valid",
  INVALID: "inValid",
  DEFAULT: null,
});

export default Vue.extend({
  data() {
    return {
      selectedFile: "",
      validationStatus: status.DEFAULT,
      description: "",
      payload: [],
      status,
      countryStatuses: [],
      importedToServer: false,
    };
  },
  methods: {
    uploadFile(event) {
      common.showLoading();
      const self = this;
      const files = event.target.files;
      this.selectedFile = event.target.files[0].name;
      Papa.parse(files[0], {
        worker: true,
        header: true,
        complete: function ({ data }) {
          if (!data.length) {
            self.wrongData = true;
          } else {
            for (let i = 0; i < data.length; i++) {
              validateFields(data[i])
                .then((response) => {
                  self.payload.push(generatePayloadFromParsedJSON(response));
                  self.validationStatus = status.VALID;
                })
                .catch((error) => {
                  self.description =
                    "On row " + (i + 1) + " " + error.toString();
                  self.validationStatus = status.INVALID;
                })
                .finally(() => {
                  common.hideLoading();
                });
            }
          }
        },
      });
    },
    submitData() {
      common.showLoading();
      const url = "/api/bff/countries/submit";
      axios
        .post(url, {
          gdhiQuestionnaires: this.payload,
        })
        .then(({ data: { countryStatuses } }) => {
          this.importedToServer = true;
          this.countryStatuses = countryStatuses;
        })
        .catch(() => {
          this.message = "Error While Importing Data To Server ";
          this.notifier({
            group: "custom-template",
            title: "Error",
            text: this.message,
            type: "error",
          });
        })
        .finally(() => {
          common.hideLoading();
        });
    },
    notifier(props) {
      this.$notify({
        group: props.group,
        title: props.title,
        text: props.text,
        type: props.type,
      });
    },
  },
  name: "UploadCSV",
});
</script>

<template>
  <div>
    <div class="download-section">
      <div class="header-bold">Template</div>
      <button
        class="btn btn-primary"
        onclick="location.href = 'https://docs.google.com/spreadsheets/d/1AZ_5oYOCokiI4xosVPCZU3jSZUYjt4MB/edit?usp=sharing&ouid=105807527850921178070&rtpof=true&sd=true';"
      >
        <i class="fa fa-download" aria-hidden="true"></i>Download
      </button>
    </div>
    <div class="header-bold">Upload File</div>
    <div class="file-name-and-error">
      <p>{{ selectedFile }}</p>
      <p class="error-message" v-if="validationStatus === status.INVALID">
        {{ description }}
      </p>
    </div>
    <div class="button-section">
      <input
        type="file"
        @change="uploadFile"
        width="100"
        height="100"
        accept="text/csv"
        class="upload-file"
      />
      <button
        class="btn btn-primary"
        onclick="document.getElementsByClassName('upload-file')[0].click();"
      >
        Select File
      </button>
      <button
        class="btn btn-primary"
        :class="
          validationStatus === status.INVALID ||
          validationStatus === status.DEFAULT
            ? 'disabled '
            : ''
        "
        :disabled="
          validationStatus === status.INVALID ||
          validationStatus === status.DEFAULT
        "
        @click="submitData"
        data-testid="import-button"
      >
        Import to server
      </button>
    </div>
    <div class="header-bold" v-if="importedToServer">Import Status</div>
    <table v-if="importedToServer" class="import-status-table">
      <thead>
        <tr>
          <th>Country</th>
          <th>Success</th>
          <th>Status</th>
          <th>Failure Reason</th>
        </tr>
      </thead>
      <tbody>
        <tr v-for="(countryStatus, id) in countryStatuses" :key="id">
          <td v-for="(countryDetails, id) in countryStatus" :key="id">
            {{ countryDetails }}
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</template>

<style scoped lang="scss">
@import "upload-csv";
</style>
