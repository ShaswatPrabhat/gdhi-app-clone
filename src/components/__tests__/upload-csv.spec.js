import { mount } from "@vue/test-utils";
import { describe, beforeEach, it, expect, vi } from "vitest";
import UploadCSV from "../uploadCSV/upload-csv.vue";
import axios from "axios";
import Papa from "papaparse";
import flushPromises from "flush-promises";
import * as uploadUtils from "../uploadCSV/uploadUtils";
import sinon from "sinon";
const event = {
  target: {
    files: [
      {
        name: "file.csv",
        size: 2,
        type: "text/csv",
      },
    ],
  },
};

describe("Upload CSV", () => {
  let wrapper;
  let countryStatuses = [
    {
      countryName: "IND",
      success: "false",
      currentStatus: "REVIEW_PENDING",
      message: "Country is already in REVIEW_PENDING state",
    },
    {
      countryName: "THA",
      success: "true",
      currentStatus: "REVIEW_PENDING",
      message: "",
    },
    {
      countryName: "PAK",
      success: "false",
      currentStatus: "PUBLISHED",
      message: "Country is already in PUBLISHED state",
    },
  ];

  const axiosPostSpy = vi.spyOn(axios, "post");
  const papaParseSpy = vi.spyOn(Papa, "parse").mockImplementation(() => null);
  const validateFieldsSpy = vi.spyOn(uploadUtils, "validateFields");
  const generatePayloadSpy = vi.spyOn(
    uploadUtils,
    "generatePayloadFromParsedJSON"
  );

  beforeEach(() => {
    axiosPostSpy.mockReset();
    papaParseSpy.mockReset();
    validateFieldsSpy.mockReset();
    generatePayloadSpy.mockReset();
    wrapper = mount(UploadCSV);
  });

  it("should submit data when the csv is uploaded", async () => {
    axiosPostSpy.mockResolvedValue({ data: countryStatuses });
    wrapper.vm.submitData();
    await flushPromises();

    expect(axiosPostSpy.mock.calls[0][0]).to.equal("/api/bff/countries/submit");
  });

  it("should not submit data when the wrong csv is uploaded and notifier should be displayed", async () => {
    let notifier = sinon.spy();

    wrapper.vm.notifier = notifier;
    axiosPostSpy.mockRejectedValue({});
    wrapper.vm.submitData();
    await flushPromises();

    sinon.assert.calledWith(notifier, {
      group: "custom-template",
      title: "Error",
      text: "Error While Importing Data To Server ",
      type: "error",
    });
  });

  it("should render upload csv", async () => {
    expect(wrapper.vm.$el).toMatchSnapshot();
  });

  it("import to server button should be disabled at start", async () => {
    expect(wrapper.vm.validationStatus).toBe("DEFAULT");
    const importButton = wrapper.find('[data-testid="import-button"]');
    expect(importButton.attributes().disabled).toEqual("disabled");
  });

  it("should call validateFields method when uploadFile is triggered", async () => {
    validateFieldsSpy.mockResolvedValue({ data: "a" });
    wrapper.vm.uploadFile(event);
    expect(papaParseSpy).toHaveBeenCalledWith(
      {
        name: "file.csv",
        size: 2,
        type: "text/csv",
      },
      expect.objectContaining({
        header: true,
        worker: true,
      })
    );
    papaParseSpy.mock.calls[0][1].complete({ data: ["abc"] });
    expect(validateFieldsSpy).toHaveBeenCalledWith("abc");
  });

  it("should set validation status to valid when validateFields is success", async () => {
    validateFieldsSpy.mockResolvedValue({ data: "a" });
    wrapper.vm.uploadFile(event);
    papaParseSpy.mock.calls[0][1].complete({ data: ["abc"] });
    await flushPromises();
    expect(generatePayloadSpy).toHaveBeenCalledOnce();
    expect(generatePayloadSpy).toHaveBeenCalledWith({ data: "a" });
    expect(wrapper.vm.validationStatus).toBe("VALID");
  });

  it("should set validation status to inValid when validateFields is failed", async () => {
    validateFieldsSpy.mockRejectedValueOnce({
      error: "error",
      value: { "Country Name": "India" },
    });
    wrapper.vm.uploadFile(event);
    papaParseSpy.mock.calls[0][1].complete({
      data: ["a"],
    });
    await flushPromises();
    expect(wrapper.vm.validationStatus).toBe("INVALID");
  });

  it("should set validation status to inValid when validateFields failed for one row and resolves for all other rows", async () => {
    validateFieldsSpy
      .mockRejectedValueOnce({
        error: "error",
        value: { "Country Name": "India" },
      })
      .mockRejectedValueOnce({
        error: "error",
        value: { "Country Name": "Germany" },
      })
      .mockResolvedValue({ data: "a" });
    wrapper.vm.uploadFile(event);
    papaParseSpy.mock.calls[0][1].complete({
      data: ["a", "b", "c"],
    });
    await flushPromises();
    expect(wrapper.vm.validationStatus).toBe("INVALID");
    expect(wrapper.vm.description).toContain("India");
  });

  it("should set validation status to invalid when file is empty", async () => {
    wrapper.vm.uploadFile(event);
    papaParseSpy.mock.calls[0][1].complete({ data: [] });
    expect(wrapper.vm.validationStatus).toBe("INVALID");
    expect(wrapper.vm.description).toBe("Empty csv");
  });
});
