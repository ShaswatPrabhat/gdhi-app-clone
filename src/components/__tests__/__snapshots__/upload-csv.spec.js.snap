// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Upload CSV > should render upload csv 1`] = `
<div
  data-v-901d73fd=""
>
  <div
    class="download-section"
    data-v-901d73fd=""
  >
    <div
      class="header-bold"
      data-v-901d73fd=""
    >
      Template
    </div>
    <button
      class="btn btn-primary"
      data-v-901d73fd=""
      onclick="location.href = 'https://docs.google.com/spreadsheets/d/1AZ_5oYOCokiI4xosVPCZU3jSZUYjt4MB/edit?usp=sharing&ouid=105807527850921178070&rtpof=true&sd=true';"
    >
      <i
        aria-hidden="true"
        class="fa fa-download"
        data-v-901d73fd=""
      />
      Download 
    </button>
  </div>
  <div
    class="header-bold"
    data-v-901d73fd=""
  >
    Upload File
  </div>
  <div
    class="file-name-and-error"
    data-v-901d73fd=""
  >
    <p
      data-v-901d73fd=""
    >
      
    </p>
    <!---->
  </div>
  <div
    class="button-section"
    data-v-901d73fd=""
  >
    <input
      accept="text/csv"
      class="upload-file"
      data-v-901d73fd=""
      height="100"
      type="file"
      width="100"
    />
    <button
      class="btn btn-primary"
      data-v-901d73fd=""
      onclick="document.getElementsByClassName('upload-file')[0].click();"
    >
       Select File 
    </button>
    <button
      class="btn btn-primary disabled "
      data-testid="import-button"
      data-v-901d73fd=""
      disabled="disabled"
    >
       Import to server 
    </button>
  </div>
  <!---->
  <!---->
</div>
`;
